src/tools/chip-cert/Cmd_PrintCD.cpp

```diff
@@ -0,0 +1,277 @@
+    kCDFormat_Base64,
+};
+
+CDFormat DetectCDFormat(const uint8_t * cd, uint32_t cdLen)
+{
+    static const uint8_t cdRawPrefix1[] = { 0x30, 0x81 };
+    static const uint8_t cdRawPrefix2[] = { 0x30, 0x82 };
+    static const char * cdHexPrefix     = "308";
+    static const char * cdB64Prefix     = "MI";
+
+    VerifyOrReturnError(cd != nullptr, kCDFormat_Unknown);
+
+    if ((cdLen > sizeof(cdRawPrefix1)) &&
+        (memcmp(cd, cdRawPrefix1, sizeof(cdRawPrefix1)) == 0 || memcmp(cd, cdRawPrefix2, sizeof(cdRawPrefix2)) == 0))
+    {
+        return kCDFormat_Raw;
+    }
+
+    if ((cdLen > strlen(cdHexPrefix)) && (memcmp(cd, cdHexPrefix, strlen(cdHexPrefix)) == 0))
+    {
+        return kCDFormat_Hex;
+    }
+
+    if ((cdLen > strlen(cdB64Prefix)) && (memcmp(cd, cdB64Prefix, strlen(cdB64Prefix)) == 0))
+    {
+        return kCDFormat_Base64;
+    }
+
+    return kCDFormat_Unknown;
+}
+
+bool ReadCD(const char * fileNameOrStr, MutableByteSpan cd)
+{
+    CDFormat cdFmt = kCDFormat_Unknown;
+    uint32_t cdLen = 0;
+    std::unique_ptr<uint8_t[]> cdBuf;
+
+    // If fileNameOrStr is a file name
+    if (access(fileNameOrStr, R_OK) == 0)
+    {
+        VerifyOrReturnError(ReadFileIntoMem(fileNameOrStr, nullptr, cdLen), false);
+
+        cdBuf = std::unique_ptr<uint8_t[]>(new uint8_t[cdLen]);
+
+        VerifyOrReturnError(ReadFileIntoMem(fileNameOrStr, cdBuf.get(), cdLen), false);
+
+        cdFmt = DetectCDFormat(cdBuf.get(), cdLen);
+        if (cdFmt == kCDFormat_Unknown)
+        {
+            fprintf(stderr, "Unrecognized CD Format in File: %s\n", fileNameOrStr);
+            return false;
+        }
+    }
+    // Otherwise, treat fileNameOrStr as a pointer to the certificate string
+    else
+    {
+        cdLen = static_cast<uint32_t>(strlen(fileNameOrStr));
+
+        cdFmt = DetectCDFormat(reinterpret_cast<const uint8_t *>(fileNameOrStr), cdLen);
+        if (cdFmt == kCDFormat_Unknown)
+        {
+            fprintf(stderr, "Unrecognized CD Format in the Input Argument: %s\n", fileNameOrStr);
+            return false;
+        }
+
+        cdBuf = std::unique_ptr<uint8_t[]>(new uint8_t[cdLen]);
+        memcpy(cdBuf.get(), fileNameOrStr, cdLen);
+    }
+
+    if (cdFmt == kCDFormat_Hex)
+    {
+        size_t len = chip::Encoding::HexToBytes(Uint8::to_char(cdBuf.get()), cdLen, cdBuf.get(), cdLen);
+        VerifyOrReturnError(CanCastTo<uint32_t>(2 * len), false);
+        VerifyOrReturnError(2 * len == cdLen, false);
+        cdLen = static_cast<uint32_t>(len);
+    }
+    else if (cdFmt == kCDFormat_Base64)
+    {
+        VerifyOrReturnError(Base64Decode(cdBuf.get(), cdLen, cdBuf.get(), cdLen, cdLen), false);
+    }
+
+    VerifyOrReturnError(cdLen <= cd.size(), false);
+    memcpy(cd.data(), cdBuf.get(), cdLen);
+
+    cd.reduce_size(cdLen);
+
+    return true;
+}
+
+void ENFORCE_FORMAT(1, 2) SimpleDumpWriter(const char * aFormat, ...)
+{
+    va_list args;
+
+    va_start(args, aFormat);
+
+    vprintf(aFormat, args);
```

Comment:
This only works on stdout, not to an output file.

Suggest have a `gDumpWriteOutputFile` in anonymous namespace and using `vfprintf`

Reply:
Added `gOutFile` variable and using `vfprintf` as suggested